### PR TITLE
Add `SPPF()` layer

### PR DESCRIPTION
**Pull request by glenn-jocher**
Implements a new cascaded SPP method I came up with today that produces mathematically identical results to SPP with less FLOPs and faster speeds :)

Can be used in model yamls to replace SPP:
```yaml
   [-1, 1, SPP, [1024, [5, 9, 13]]],  # original
   [-1, 1, SPPF, [1024, 5]],  # replacement (equivalent to above)
```
or
```yaml
   [-1, 1, SPP, [1024, [3, 5, 7]]],  # original
   [-1, 1, SPPF, [1024, 3]],  # replacement (equivalent to above)
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to YOLOv5 with a new Spatial Pyramid Pooling - Fast (SPPF) layer and profiling option.

### 📊 Key Changes
- Introduced the new `SPPF` layer as an alternative to the standard `SPP` layer in Spatial Pyramid Pooling.
- Added comments providing a reference to the academic paper on SPP.
- Included the `SPPF` layer in the model configuration options.
- Integrated a `--profile` argument to measure model inference speed.

### 🎯 Purpose & Impact
- **SPPF Implementation**: Offers a faster variant of the original SPP layer, which may improve inference speeds with potentially minimal impact on accuracy.
- **Research Integration**: Citing the SPP academic paper helps in understanding the background of the technology.
- **Configuration Flexibility**: Users can now choose between using `SPP` or `SPPF` in their models for potentially better performance.
- **Performance Measurement**: The `--profile` flag allows developers and researchers to measure and optimize the speed of their models.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Profiling results:
```python

# Profile
from utils.torch_utils import profile
from models.common import SPP, SPPF

m1 = SPP(1024, 1024)
m2 = SPPF(1024, 1024)
results = profile(input=torch.randn(16, 1024, 64, 64), ops=[m1, m2], n=100)
```

Screenshot 2021-08-15 at 14 19 50

**Comment by glenn-jocher** [comment]
allclose verification

```python
import torch
import torch.nn as nn

m5 = nn.MaxPool2d(kernel_size=13, stride=1, padding=13 // 2)
m3 = nn.MaxPool2d(kernel_size=5, stride=1, padding=5 // 2)

x = torch.randn(1, 3, 64, 64)
y5 = m5(x)
y3 = m3(m3(m3(x)))

print(y5)
print(y3)

print(torch.allclose(y5,y3))
```

**Comment by MAO-S-K** [comment]
Hello, I would like to ask what is the internal structure of SPPF?

**Comment by glenn-jocher** [comment]
@15171461072 see https://github.com/ultralytics/yolov5/pull/4420/files

**Comment by ehdrndd** [comment]
i tested entire module examples using same convolution weights.
result : !same!
```Python
import torch
import torch.nn as nn

def autopad(k, p=None):  # kernel, padding
    # Pad to 'same'
    if p is None:
        p = k // 2 if isinstance(k, int) else [x // 2 for x in k]  # auto-pad
    return p

class Conv(nn.Module):
    # Standard convolution
    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, act=True):  # ch_in, ch_out, kernel, stride, padding, groups
        super().__init__()
        self.conv = nn.Conv2d(c1, c2, k, s, autopad(k, p), groups=g, bias=False)
        self.bn = nn.BatchNorm2d(c2)
        self.act = nn.SiLU() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())

    def forward(self, x):
        return self.act(self.bn(self.conv(x)))

    def forward_fuse(self, x):
        return self.act(self.conv(x))

class SPP(nn.Module):
    # Spatial Pyramid Pooling (SPP) layer https://arxiv.org/abs/1406.4729
    def __init__(self, c1, c2, k=(5, 9, 13)):
        super().__init__()
        c_ = c1 // 2  # hidden channels
        self.cv1 = Conv(c1, c_, 1, 1)
        self.cv2 = Conv(c_ * (len(k) + 1), c2, 1, 1)
        self.m = nn.ModuleList([nn.MaxPool2d(kernel_size=x, stride=1, padding=x // 2) for x in k])

    def forward(self, x):
        x = self.cv1(x)
        return self.cv2(torch.cat([x] + [m(x) for m in self.m], 1))


class SPPF(nn.Module):
    # Spatial Pyramid Pooling - Fast (SPPF) layer for YOLOv5 by Glenn Jocher
    def __init__(self, c1, c2, k=5):  # equivalent to SPP(k=(5, 9, 13))
        super().__init__()
        c_ = c1 // 2  # hidden channels
        self.cv1 = Conv(c1, c_, 1, 1)
        self.cv2 = Conv(c_ * 4, c2, 1, 1)
        self.m = nn.MaxPool2d(kernel_size=k, stride=1, padding=k // 2)

    def forward(self, x):
        x = self.cv1(x)
        y1 = self.m(x)
        y2 = self.m(y1)
        return self.cv2(torch.cat([x, y1, y2, self.m(y2)], 1))

spp = SPP(32, 32, k=(5,9,13))
sppf = SPPF(32, 32, k=5)

sppf.cv1 = spp.cv1
sppf.cv2 = spp.cv2

x = torch.randn(1, 32, 64, 64)
y5 = spp(x)
y3 = sppf(x)

print(y5)
print(y3)

print(torch.allclose(y5,y3))
```

**Comment by glenn-jocher** [comment]
@ehdrndd awesome! Thanks for the validation results :)

**Comment by K-tang-mkv** [comment]
@glenn-jocher Why do we need sppf or spp layer in yolov5 network? As I know, the spp layer is to solve the multi-size image inputs so that the output vector size after spp layer is fixed for the next fully connected layer. But for the yolov5 net, the next layer after sppf layer is conv layer, which can accept any input with different sizes. I really appreciate any help you can provide.

**Comment by glenn-jocher** [comment]
@K-tang-mkv we don't 'need' any layer in YOLOv5. They all work together to incrementally improve detection results.

**Comment by K-tang-mkv** [comment]
> @K-tang-mkv we don't 'need' any layer in YOLOv5. They all work together to incrementally improve detection results.

@glenn-jocher Oh, I got it. Thank you so much.

**Comment by glenn-jocher** [comment]
@K-tang-mkv you're welcome! If you have any more questions or need further assistance, feel free to ask. Good luck with your project!
